### ISSUE TITLE
Not working on m1 mac

### ISSUE THREAD

**Issue by skorphil** [xzhih/one-key-hidpi#207]
Seems like its not working with m1. I used script on my old macbook (intel) and it worked fine. On m1 macbook after rebooting i do not see any results. Tried with and without EDID

**Comment by wey-gu** [comment]
+1
Confirmed it's not working on the M1 MacBook on a 2K monitor.
macOS Big Sur 11.6

**Comment by smoky96** [comment]
+1，m1 macOS Big Sur 11.6，这条命令：ioreg -lw0 | grep -i "IODisplayEDID" 的结果是空的

**Comment by Tongzhao9417** [comment]
前面的issues里有提到过了，2k分辨率在m1下不生效的，现在暂时无解

**Comment by skorphil** [comment]
I found out that there is no solution for m1 macs because apple changed something with puprpose. So m1 basically not support hdpi on monitors less than 4k

**Comment by ppkantorski** [comment]
idk how but i seem to have gotten HiDPi working on my m1 mac mini
![Screen Shot 2021-11-03 at 10 53 45 PM](https://user-images.githubusercontent.com/6467366/140269009-85dd8c70-38a2-4a48-a7b3-be420b2c6e8b.png)

**Comment by waydabber** [comment]
Nice. But I think you got it working because 4K displays should be by default supported for HiDPI by M1 macs and something was/is messed up somehow somewhere which prevented this in your config maybe even though you use a 4K display.

But if you found a reliable way to circumvent this current M1 limitation without a dummy on sub-4K displays, it would be great if you shared it with the community in a way that could be reproduced. The BetterDummy workaround is less than ideal honestly and it would be great if native support could be achieved. 🙂

**Comment by HQQ123** [comment]
This problem is too annoying!

**Comment by wey-gu** [comment]
> This problem is too annoying!

https://github.com/waydabber/BetterDummy works fine for M1 Mac.

**Comment by ibehnam** [comment]
Switchresx works too.

**Comment by ibehnam** [comment]
See this for a thorough info:

https://medium.com/@parttimeben/mac-it-just-works-horribly-c98fd9be4001

**Comment by waydabber** [comment]
Yes, if you have any Retina compatible display already connected, you can use that to mirror its screen to your non-Retina display for HiDPI. SwitchResX will let you create native resolutions on M1 (but can't activate HiDPI) but if you create a native resolution on a display that already has HiDPI created, you can match the aspect ratio of your external screen. The obvious disadvantage of this method is that you always need a separate display occupied only to allow mirroring. On Macs with multiple outputs you can use a physical dummy to circumvent this which mostly works (the drawbacks are sometimes flickering mouse and the possibility that the login screen will show on your physical dummy screen if mirroring fails which will require you to disconnect manually - but to be fair, BetterDummy has its headaches as well :)).

**Comment by waydabber** [comment]
Hey @ibehnam - I think neither SwitchResX nor BetterDummy should be that heavy on resources as in case of mirroring both approaches use an intra-GPU, hardware mirror which means that there should be a single common framebuffer (but even if the framebuffer is copied, everything happens on-chip for M1 machines and the framebuffers are not too large big actually, the desktop compositing - which might be the heavier thing - happens only once in both scenarios).

BetterDummy can do streaming instead of mirroring, that will actually have two desktops, but the actual process is also super-efficient (the source screen is applied to a window layer surface which is optimized at hardware/kernel level to work ultra fast on M1 machines).

On Intel you'll have some issues with BetterDummy especially with non-integrated GPUs as in that scenario a lot of data is being moved around but on Intel SwitchResX is unquestionably the best choice to create custom resolutions.

BetterDummy has some issues on some configs with mirroring which are related to various macOS bugs when mirroring a virtual display (this happens with Sidecar as well without BetterDummy) - there can be some sleep woes and sometimes flickering when there is some screen activity (for example mouse is moved around) due to some weird colorsync problems. This can be annoying and require disconnecting/reconnecting the dummy or turn off/on mirroring.

**Comment by ibehnam** [comment]
@waydabber thanks for the details. I followed [this guide](https://contra.medium.com/how-to-turn-on-hidpi-mode-on-external-display-in-macos-5c1035dc42b5) and [this one](https://codeclou.github.io/Display-Override-PropertyList-File-Parser-and-Generator-with-HiDPI-Support-For-Scaled-Resolutions/) to create `*.plist` files similar to what Switchresx does, but unfortunately, RDM doesn't pick them up. Given that `/System/Library` is read-only on latest macOS versions, I can only guess that Switchresx copies such files to `/Library` instead. But on my M1 MBP, this method no longer works.

So, if HiDPI resolution are [disabled by Apple](https://www.madrau.com/support/support/faq_files/ns_Why_isnt_there_any_HiDPI_resolu.html), I don't know how Switchresx manages to create them. Really curious!

I had a better experience with BetterDummy. But when using MonitorControl + BetterDummy, sometimes the external display starts flickering and the only way to fix it it to reconnect the HDMI cable.

**Comment by waydabber** [comment]
Hi @ibehnam - yes, afaik SwitchResX can't create HiDPI resolutions on displays where this feature is disabled by Apple so in this regard there is no alternative currently to BetterDummy. SwitchResX can create custom scaled resolutions though on Apple Silicon. If HiDPI is already enabled for a display, then the relevant HiDPI versions of the created custom scaled resolutions should be accessible as well.

I know the mirror flickering issue is a pain. :(

**Comment by ibehnam** [comment]
@waydabber So the way I got Switchresx working was by creating the HiDPI resolution on the _internal_ display, and then mirroring it to the external. Obviously, the downside is that two displays must be on and one can't close the lid in this case, which is why I prefer BetterDummy.
To create a HiDPI resolution on Switchresx, I simply created a 3840x2160 (2x the 1920x1080 resolution). After restart, my custom resolution showed up as 1920x1080 HiDPI!

I hope the flickering issue goes away, but tbh I think it's more likely that Apple will break something in the next macOS :/

**Comment by waydabber** [comment]
Yes, Apple breaking stuff is a possibility anytime - especially since all these solutions are using various undocumented features or stuff that was not envisioned originally to be used for such purposes. But I don't think Apple will intentionally cripple or remove stuff (macOS is literally littered with old stuff inside) - they usually break things only when they don't consider a current thing a documented feature and it gets in the way of something new/better. But messing with virtual screens in a bad way would cause a lot of issues (render DisplayLink not working for example which would restart the DisplayLink support saga that happened a few years ago after the driver model changes) so I don't think such a thing will happen. We'll see.
